### PR TITLE
Report diagnostic counters as Delta Counters

### PR DESCRIPTION
**Pull request by joannak-vmware**
* Report ~sdk diagnostic counters as Delta Counters

* Created WavefrontSdkDeltaCounter class

* Replaced all existing internal WavefrontSdkCounters with WavefrontSdkDeltaCounters

* Updated correlating Counter functions

### CODE REVIEW

**Comment by joannak-vmware** [comment]
Am working on failed build.

File: src/main/java/com/wavefront/sdk/common/metrics/WavefrontSdkCounter.java [hunk 1]
```diff
@@ -30,6 +30,22 @@ public void inc(long n) {
...
+    dec(1);
+  }
+
+  /**
+   * Decrements the counter by the specified amount.
+   *
+   * @param n The amount to decrement by.
+   */
+  public void dec(long n) {
+    count.addAndGet(-1 * n);
```
Comment (hanwavefront) [inline src/main/java/com/wavefront/sdk/common/metrics/WavefrontSdkCounter.java:46]: nit: can be `-n`

File: src/main/java/com/wavefront/sdk/common/metrics/WavefrontSdkMetricsRegistry.java [hunk 2]
```diff
@@ -196,10 +206,21 @@ public void close() {
    * @param name  The metric name.
    * @return A new or pre-existing counter.
    */
-  public WavefrontSdkCounter newCounter(String name) {
+  public WavefrontSdkCounter newCounter (String name) {
```
Comment (hanwavefront) [inline src/main/java/com/wavefront/sdk/common/metrics/WavefrontSdkMetricsRegistry.java:209]: nit: space between newCounter and (

File: src/test/java/com/wavefront/sdk/common/metrics/WavefrontSdkMetricsRegistryTest.java [hunk 3]
```diff
@@ -42,4 +42,25 @@ public void testCounter() {
...
+    assertEquals(3, deltaCounter.count());
+    deltaCounter.dec();
+    assertEquals(2, deltaCounter.count());
+    deltaCounter.dec(deltaCounter.count());
+    assertEquals(0, deltaCounter.count());
+
+    //Delta Counters decrements counters each time after data is sent. New counters with same name will have 0 count.
+    assertEquals(0, registry.newDeltaCounter("deltaCounter").count());
+    assertEquals(0, registry.newDeltaCounter("deltaCounter2").count());
+  }
```
Comment (hanwavefront) [inline src/test/java/com/wavefront/sdk/common/metrics/WavefrontSdkMetricsRegistryTest.java:65]: maybe also add a test to check that the value of the delta is reset after it is sent.

File: src/main/java/com/wavefront/sdk/common/Utils.java [hunk 4]
```diff
@@ -454,11 +454,16 @@ private static String sanitizeInternal(String s, boolean addQuotes) {
     for (int i = 0; i < s.length(); i++) {
       char cur = s.charAt(i);
       boolean isLegal = true;
+      boolean isTildaPrefixed = s.charAt(0) == 126;
+      boolean isDeltaPrefixed = (s.charAt(0) == 0x2206) || (s.charAt(0) == 0x0394);
+      boolean isDeltaTildaPrefixed = isDeltaPrefixed && s.charAt(1) == 126;
       if (!(44 <= cur && cur <= 57) && !(65 <= cur && cur <= 90) && !(97 <= cur && cur <= 122) &&
           cur != 95) {
-        if (!((i == 0 && cur == 0x2206) || (i == 0 && cur == 0x0394) || (i == 0 && cur == 126))) {
+        if (!(i == 0 && (isDeltaPrefixed || isTildaPrefixed) || (i == 1 && isDeltaTildaPrefixed))) {
```
Comment (hanwavefront) [inline src/main/java/com/wavefront/sdk/common/Utils.java:462]: :thumbsup:

**Review by hanwavefront — CHANGES_REQUESTED** [review]
a couple minor things, LGTM otherwise

**Review by srujann — CHANGES_REQUESTED** [review]
minor comments.

**Review by hanwavefront — APPROVED** [review]
LGTM, thanks

**Review by hanwavefront — APPROVED** [review]

File: src/main/java/com/wavefront/sdk/common/metrics/WavefrontSdkMetricsRegistry.java [hunk 5]
```diff
@@ -159,6 +160,15 @@ public void run() {
             Number value = (Number) ((WavefrontSdkGauge) metric).getValue();
             wavefrontMetricSender.sendMetric(name, value.doubleValue(), timestamp, source, tags);
           }
+        }  else if (metric instanceof WavefrontSdkDeltaCounter) {
+          if (!name.startsWith(Constants.DELTA_PREFIX) && !name.startsWith(Constants.DELTA_PREFIX_2)) {
+            name = Constants.DELTA_PREFIX + name;
+          }
```
Comment (srujann) [inline src/main/java/com/wavefront/sdk/common/metrics/WavefrontSdkMetricsRegistry.java:166]: why can't we do this when registering the metric into the `metrics` CHM?

Comment (joannak-vmware) [reply src/main/java/com/wavefront/sdk/common/metrics/WavefrontSdkMetricsRegistry.java]: `metrics` do not have full name, so we add the Delta to the name before sending.

**Review by srujann — APPROVED** [review]

**Review by sushantdewan123 — APPROVED** [review]

**Pull request merged**